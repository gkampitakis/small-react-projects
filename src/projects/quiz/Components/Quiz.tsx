import React, { ReactElement } from 'react';
import Loading from '../../../components/Loading';
import SetupForm from './SetupForm';
import Modal from './Modal';
import { useGlobalContext } from '../context';

export default function Quiz (): ReactElement {
  const { waiting, loading, questions, currentQ, correct, nextQuestion, checkAnswer } = useGlobalContext();
  if (waiting) {
    return <SetupForm />;
  }

  if (loading) {
    return <Loading />;
  }

  const { incorrect_answers, correct_answer, question } = questions[currentQ];

  const answers = [...incorrect_answers];

  const tempIndex = Math.floor(Math.random() * 4)
  if (tempIndex === 3) {
    answers.push(correct_answer)
  } else {
    answers.push(answers[tempIndex])
    answers[tempIndex] = correct_answer
  }

  return (
    <main className="quiz_project">
      <Modal />
      <section className="quiz">
        <p className="correct-answers">correct answers : {correct}/{currentQ}</p>
        <h2>{question}</h2>
        <div className="btn-container">
          {answers.map((a, i) => <button onClick={() => checkAnswer(correct_answer === a)} className="answer-btn" key={i}>{a}</button>)}
        </div>
        <button className="next-question" onClick={nextQuestion}>
          next question
        </button>
      </section>
    </main>
  );
}
