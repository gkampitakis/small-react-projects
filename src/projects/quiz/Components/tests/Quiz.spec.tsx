import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Quiz from '../Quiz';

jest.mock('../../context', () => {
  let isWaiting = true;
  let isLoading = true;
  let mockError = false;
  const nextSpy = jest.fn();
  const setUpFormOnChangeSpy = jest.fn();
  const setupFormOnSubmitSpy = jest.fn();

  return {
    useGlobalContext: () => ({
      waiting: isWaiting,
      loading: isLoading,
      currentQ: 0,
      correct: 0,
      isModalOpen: false,
      error: mockError,
      quiz: {
        amount: 10,
        category: 'sports',
        difficulty: 'easy'
      },
      nextQuestion: nextSpy,
      setupFormOnChange: setUpFormOnChangeSpy,
      setupFormOnSubmit: (e: Event) => {
        e.preventDefault();
        setupFormOnSubmitSpy();
      },
      checkAnswer: () => true,
      questions: [
        {
          incorrect_answers: ['a', 'a', 'a'],
          correct_answer: 'a',
          question: 'question'
        }
      ]
    }),
    mockWaiting: (value: boolean) => (isWaiting = value),
    mockLoading: (value: boolean) => (isLoading = value),
    nextSpy,
    setupFormOnSubmitSpy,
    setUpFormOnChangeSpy,
    mockError: (value: boolean) => (mockError = value)
  };
});

describe('Quiz', () => {
  const {
    setUpFormOnChangeSpy,
    setupFormOnSubmitSpy,
    mockWaiting,
    mockLoading,
    mockError
  } = jest.requireMock('../../context');

  beforeEach(() => {
    setUpFormOnChangeSpy.mockClear();
    setupFormOnSubmitSpy.mockClear();
    mockWaiting(false);
    mockLoading(false);
    mockError(false);
  });

  it('Should render correctly', () => {
    const tree = renderer.create(<Quiz />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('Setup Form', () => {
    beforeEach(() => {
      mockWaiting(true);
    });

    it('Should call onChange function', () => {
      render(<Quiz />);

      userEvent.selectOptions(screen.getByLabelText(/category/i), 'politics');
      userEvent.selectOptions(screen.getByLabelText(/difficulty/i), 'medium');
      userEvent.type(screen.getByLabelText(/number of questions/i), '2');

      expect(setUpFormOnChangeSpy).toHaveBeenCalledTimes(3);
    });

    it('Should call onSubmit function', () => {
      render(<Quiz />);

      userEvent.click(screen.getByText(/start/i));

      expect(setupFormOnSubmitSpy).toHaveBeenCalledTimes(1);
    });

    it('Should show error message', () => {
      mockError(true);

      const { baseElement } = render(<Quiz />);

      expect(baseElement).toMatchSnapshot();
    });
  });

  it('Should show loading', () => {
    mockLoading(true);

    const { baseElement } = render(<Quiz />);

    expect(baseElement).toMatchSnapshot();
  });
});
