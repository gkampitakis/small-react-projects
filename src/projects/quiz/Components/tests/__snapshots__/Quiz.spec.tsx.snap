// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz Setup Form Should show error message 1`] = `
<body>
  <div>
    <main
      class="quiz_project"
    >
      <form
        class="quiz setup_form"
      >
        <h2>
          Setup Quiz
        </h2>
        <div
          class="form-control"
        >
          <label
            for="amount"
          >
            Number of questions
          </label>
          <input
            autocomplete="off"
            class="form-input"
            id="amount"
            max="50"
            min="1"
            name="amount"
            type="number"
            value="10"
          />
        </div>
        <div
          class="form-control"
        >
          <label
            for="category"
          >
            category
          </label>
          <select
            class="form-input"
            id="category"
            name="category"
          >
            <option
              value="sports"
            >
              sports
            </option>
            <option
              value="history"
            >
              history
            </option>
            <option
              value="politics"
            >
              politics
            </option>
          </select>
        </div>
        <div
          class="form-control"
        >
          <label
            for="difficulty"
          >
            Select Difficulty
          </label>
          <select
            class="form-input"
            id="difficulty"
            name="difficulty"
          >
            <option
              value="easy"
            >
              easy
            </option>
            <option
              value="medium"
            >
              medium
            </option>
            <option
              value="hard"
            >
              hard
            </option>
          </select>
        </div>
        <p
          class="error"
        >
          Canâ€˜t generate questions, please try different options
        </p>
        <button
          type="submit"
        >
          Start
        </button>
      </form>
    </main>
  </div>
</body>
`;

exports[`Quiz Should render correctly 1`] = `
<main
  className="quiz_project"
>
  <div
    className="modal-container"
  >
    <div
      className="modal-content"
    >
      <h2>
        congrats!
      </h2>
      <p>
        You answered 
        0
        % of questions correctly
      </p>
      <button
        className="close-btn"
      >
        play again
      </button>
    </div>
  </div>
  <section
    className="quiz"
  >
    <p
      className="correct-answers"
    >
      correct answers : 
      0
      /
      0
    </p>
    <h2>
      question
    </h2>
    <div
      className="btn-container"
    >
      <button
        className="answer-btn"
        onClick={[Function]}
      >
        a
      </button>
      <button
        className="answer-btn"
        onClick={[Function]}
      >
        a
      </button>
      <button
        className="answer-btn"
        onClick={[Function]}
      >
        a
      </button>
      <button
        className="answer-btn"
        onClick={[Function]}
      >
        a
      </button>
    </div>
    <button
      className="next-question"
      onClick={[MockFunction]}
    >
      next question
    </button>
  </section>
</main>
`;

exports[`Quiz Should show loading 1`] = `
<body>
  <div>
    <main
      class="loading_component"
    >
      <div
        class="loader"
      >
        <div
          class="bar"
        />
      </div>
      <img
        alt="Loading Logo"
        class="logo"
        src="loading.png"
      />
    </main>
  </div>
</body>
`;
